File: ecommerce_project/src/main/java/com/ecommerce/ecommerce_project/entity/Product.java
package com.ecommerce.ecommerce_project.entity;


import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;
import lombok.*;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Entity
@Table(name = "product", indexes = {
        @Index(name = "idx_product_brand", columnList = "brand"),
        @Index(name = "idx_product_title", columnList = "title"),

})
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@ToString
@Builder
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String imageUrl;

    @Column(nullable = false)
    private String brand;

    @Column(nullable = false)
    private String title;

    @Column(nullable = false)
    private String color;

    @Column(nullable = false)
    private Double discountedPrice;

    @Column(nullable = false)
    private Double price;

    @Column(nullable = false)
    private Double discountPercent;

    @ElementCollection
    @CollectionTable(name = "product_size", joinColumns = @JoinColumn(name = "product_id"))
    @Builder.Default
    @ToString.Exclude
    private Set<Size> size = new HashSet<>();

    @Column(nullable = false)
    private Integer quantity;

    @Column(length = 5000)
    private String description;

    @JsonManagedReference(value = "product_review")
    @OneToMany(mappedBy = "product",cascade = CascadeType.ALL)
    @Builder.Default
    @ToString.Exclude
    private List<Review> reviews=new ArrayList<>();


    @ManyToOne()
    @JoinColumn(name = "product_category")
    private Category category;

    @PrePersist
    public void calculateDiscountedPrice(){
        this.discountedPrice=price-(discountPercent*price)/100.0;
        this.discountedPrice=Math.round(this.discountedPrice*100.0)/100.0;
    }


}

File: e-commerce/src/types/product.ts
export interface Category{
    id:number;
    name:string;
    level:number;
    parentCategory?:Category;
}
export interface ProductSize{
    name:string;
    quantity:number;
}

export interface Product {
    id: number;
    title: string;
    brand: string;
    color: string;
    price: number;
    discountedPrice: number;
    discountPercent: number;
    imageUrl: string;
    quantity: number;
    description?: string;
    size: ProductSize[];
    category?: Category;
}

export interface ProductResponse{
    content:Array<Product>;
    totalPages:number;
    totalElements:number;
    size:number;
    number:number;
    last:boolean;
    first:boolean;
}
export interface ProductFilters {
    category?: string;
    minPrice?: number;
    maxPrice?: number;
    colors?: string;
    sizes?: string;
    minDiscount?: number;
    sort?: string;
    page: number;
    size: number;
}